packages/astroport/src/generator.rs

```diff
@@ -186,6 +201,82 @@ pub struct PendingTokenResponse {
     pub pending_on_proxy: Option<Uint128>,
 }
 
+/// Vec wrapper for internal use.
+/// Some business logic relies on an order of this vector, thus it is forbidden to sort it
+/// or remove elements. New values can be added using .update() ONLY.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema, Default)]
+pub struct RestrictedVector<T>(Vec<(Addr, T)>);
+
+pub trait Increaseable
+where
+    Self: Sized,
+{
+    fn increase(self, new: Self) -> StdResult<Self>;
+}
+
+impl<T> RestrictedVector<T>
+where
+    T: Copy + Increaseable,
+{
+    pub fn new(first_proxy: Addr, first_reward_index: T) -> Self {
+        Self(vec![(first_proxy, first_reward_index)])
+    }
+
+    pub fn get_last_mut(&mut self, proxy: &Addr) -> StdResult<&mut T> {
+        self.0
+            .last_mut()
+            .filter(|(p, _)| p.as_str() == proxy.as_str())
+            .map(|(_, v)| v)
+            .ok_or_else(|| StdError::generic_err(format!("Proxy {} not found", proxy)))
+    }
+
+    pub fn get_last(&self, proxy: &Addr) -> StdResult<T> {
+        self.0
+            .last()
+            .filter(|(p, _)| p.as_str() == proxy.as_str())
+            .map(|(_, v)| v)
+            .cloned()
+            .ok_or_else(|| StdError::generic_err(format!("Proxy {} not found", proxy)))
+    }
+
+    pub fn update(&mut self, key: &Addr, value: T) -> StdResult<()> {
+        let proxy_ref = self
+            .0
+            .iter_mut()
+            .find(|(proxy_addr, _)| proxy_addr.as_str() == key.as_str());
+        match proxy_ref {
+            Some((ref proxy, index)) if proxy.as_str() == key.as_str() => {
```

Comment:
This is just for extra safety. Since we cannot use HashMap we cannot guarantee that the last value corresponds to a current proxy. I tried to restrict it by creating this RestrictedVector struct, but there is always exists a way to rewrite a whole vector in, for example, ```POOL.accumulated_proxy_rewards_per_share``` field.